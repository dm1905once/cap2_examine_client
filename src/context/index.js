import React from 'react';

export const QuestionDetailsContext = React.createContext();
export const AuthContext = React.createContext();