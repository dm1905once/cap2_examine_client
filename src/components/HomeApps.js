import React, { useState, useEffect, useContext } from 'react';
import { useLocation, useHistory, Redirect } from 'react-router-dom';
import { loadStripe } from "@stripe/stripe-js";
import { AuthContext } from "../context";
import examineApi from '../apis/examineApi';
import ExamList from './applicants/ExamList';
import AppAccess from './applicants/AppAccess';
import AppRecords from './applicants/AppRecords';
const stripePromise = loadStripe("pk_test_51HcDauJMaHZnra3gtM9N5ZNXiYqFIkSYKKWs5GxoG3sAtyxIUJFKaXWpbLvl37OcSa2bd03rYBlP2J0Yc8a5ZkvV00clsLscYO");

const HomeApps = () => {
    const location = useLocation();
    const history = useHistory();
    const { isApplicantAuth, userInfo } = useContext(AuthContext);
    const [ examList, setExamList ] = useState([]);
    const [ applicationDetails, setApplicationDetails ] = useState({status: 'reloaded'});
    const [ stripeSession, setStripeSession ] = useState(null);
    const [ topMessage, setTopMessage ] = useState();

    useEffect(() => {
        // Check to see if this is a redirect back from Checkout
        const query = new URLSearchParams(window.location.search);
    
        if (query.get("success")) {
          const application_id=query.get("application_id");
          const exam_id=query.get("exam_id");
          const applicant_email=query.get("applicant_email");
          registerPurchasedExam({application_id, exam_id, applicant_email});
        }
    
        if (query.get("canceled")) {
            setTopMessage({type: 'Info', message: "Order canceled -- continue to look around and take exam when you're ready."});
        }
      }, []);


    useEffect(()=>{
        async function loadStripeSession(){
            const stripe = await stripePromise;
            const result = await stripe.redirectToCheckout({
                sessionId: stripeSession.id,
            });

            if (result.error) {
                setTopMessage({type: 'Error', message: "A problem occurred while attempting to acquire the exam"});
                window.scrollTo(0,0);
            } 
        };
        if (stripeSession) loadStripeSession();
    }, [stripeSession]);


    useEffect(()=>{
        async function retrieveExamList(){
            const applicableExams = await examineApi.getApplicableExams();
            setExamList(applicableExams);
        };
        retrieveExamList();
    },[]);


    async function registerPurchasedExam(appDetails){
        const newApplication = await examineApi.acquireExam(appDetails);
        
        if (newApplication === null ) {
            history.push("/applicants");
            setTopMessage({type: 'Error', message: "A problem occurred while attempting to acquire the exam"});
        } else {
            setTopMessage({type: 'Success', message: "Thank you! The exam has been added to your list of Purchased Exams. Good Luck!"});
        }
    };


    const handleBuyExam = async (e, exam_id, application_id, exam_name, org_logo) =>{
        if (!isApplicantAuth) {
            setTopMessage({type: 'Warning', message: "Please login or register first"});
            window.scrollTo(0,0);
        } else {
            const buyExamButton = e.currentTarget;
            buyExamButton.classList.toggle("loading");

            const appDetails = {
                application_id,
                applicant_email: userInfo.email,
                exam_id,
                exam_name,
                org_logo
            };

            setApplicationDetails(appDetails);

            async function createStripeSession(){
                try {
                    setStripeSession(await examineApi.createStripeSession(appDetails));
                } catch(e){
                    buyExamButton.classList.toggle("loading");
                    setTopMessage({type: 'Error', message: "We're sorry! An error occurred while trying to acquire the exam."});
                    window.scrollTo(0,0);
                }
            };
            createStripeSession();
        }
    };

    return (
        <div className="ui container">
            {topMessage?<TopMessage content={topMessage}/>:''}

            {isApplicantAuth? <AppRecords /> : <AppAccess />}

            <h2 className="ui block header">
                <i className="tasks icon"></i>
                <div className="content">Browse exams</div>
            </h2>
            {
                examList.length === 0
                ?   <div>
                        <h2 className="ui blue header">Loading exam list</h2>
                        {[1,2,3,4,5,6,7,8,9,0].map(x=>
                            <div className="ui placeholder" key={x}>
                                <div className="image header">
                                    <div className="line"></div>
                                    <div className="line"></div>
                                    <div className="line"></div>
                                </div>
                            </div>
                        )}
 
                    </div>
                :   
                    <div className="ui cards">
                        {examList.map(exam=>
                            <ExamList exam={exam} 
                                    key={exam.exam_id}
                                    handleBuyExam={handleBuyExam} />)}
                    </div>
            }
        </div>
    )

    function TopMessage({content}){
        let color;
        switch (content.type) {
            case 'Success': color = 'green'; break;
            case 'Error': color = 'red'; break;
            case 'Info': color = 'blue'; break;
            case 'Warning': color = 'yellow'; break;
            default: color = 'yellow';
        };

        if (Array.isArray(content.message)){
            return (
                <div className={`ui message ${color}`}>
                    <div className="header">{content.type} message</div>
                    <ul className="list">
                        {content.message.map((msg,i)=><li key={i}>{msg}</li>)}
                    </ul>
                </div>
            )
        }else {
            return (<div className={`ui warning message ${color}`}>{content.message}</div>)
        }
    };

}

export default HomeApps;