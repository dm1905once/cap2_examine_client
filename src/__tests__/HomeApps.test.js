import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import uniqid from 'uniqid';
import HomeApps from '../components/HomeApps';
import { AuthContext } from "../context";

function renderHomeApps(applicantContext){
    return render(
        <AuthContext.Provider value={applicantContext}>
            <HomeApps />
        </AuthContext.Provider>
    )
}

// Smoke test
test('HomeApps renders without crashing - no applicant logged in', ()=>{
    renderHomeApps({isApplicantAuth: false, applicantInfo: undefined});
});

test('HomeApps renders without crashing - applicant logged in', ()=>{
    const applicantContext = {
        isApplicantAuth: true, 
        applicantInfo: {
            email: "some@email.com",
            role: 'applicant'
        }};
    <BrowserRouter>renderHomeApps(applicantContext)</BrowserRouter>;
});


// Snapshot test
test('HomeApps matches snapshot', ()=>{
    const applicantContext = {isApplicantAuth: false, applicantInfo: undefined};
    const { asFragment } = renderHomeApps(applicantContext);
    expect(asFragment()).toMatchSnapshot();
});

// Specialized tests
test('Applicant registration', async ()=>{
    // Select Registration fields
    const applicantHome = renderHomeApps({isApplicantAuth: false, applicantInfo: undefined});
    const fullNameIn = applicantHome.getByPlaceholderText('First and Last Name');
    const emailIn = applicantHome.getByPlaceholderText('Email Address');
    const pwdIn1 = applicantHome.getByTestId('loginpwd1') 
    const pwdIn2 = applicantHome.getByTestId('loginpwd2') 
    const registerButton = applicantHome.getAllByRole('button')[1];
    expect(fullNameIn.toBeInTheDocument);
    expect(pwdIn1.toBeInTheDocument);
    expect(registerButton.toBeInTheDocument);

    // Populate Registration fields
    const userSuffix = uniqid.time();
    await wait(()=>{
        fireEvent.change(fullNameIn, { target: { value: `Mock_${userSuffix} Name` } });
        fireEvent.change(emailIn, { target: { value: "TestEmail" } });
        fireEvent.change(pwdIn1, { target: { value: "qwer" } });
        fireEvent.change(pwdIn2, { target: { value: "qwer" } });
    }).then(fireEvent.click(registerButton));
    expect(applicantHome.getByText("Invalid email address")).toBeInTheDocument;

    await wait(()=>{
        fireEvent.change(emailIn, { target: { value: "TestEmail@test.com" } });
    }).then(fireEvent.click(registerButton));
    expect(applicantHome.queryByText("Invalid email address")).not.toBeInTheDocument;

});