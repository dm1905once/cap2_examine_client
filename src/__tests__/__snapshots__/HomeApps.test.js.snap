// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeApps matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui container"
  >
    <div>
      <h2
        class="ui block header"
      >
        <i
          class="address card outline icon"
        />
        <div
          class="content"
        >
          Login or Register
        </div>
      </h2>
      <div
        class="ui placeholder segment"
      >
        <div
          class="ui two column very relaxed stackable grid"
        >
          <div
            class="middle aligned row"
          >
            <div
              class="column"
            >
              <h1>
                Applicant Login
              </h1>
              <div>
                <form
                  class="ui equal width form "
                >
                  <div
                    class="fields"
                  >
                    <div
                      class="field required"
                    >
                      <label>
                        Email Address
                      </label>
                      <input
                        name="email"
                        placeholder=""
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="fields"
                  >
                    <div
                      class="field required"
                    >
                      <label>
                        Password
                      </label>
                      <input
                        name="password"
                        type="password"
                        value=""
                      />
                    </div>
                  </div>
                  <button
                    class="ui primary blue right floated right labeled icon button"
                    type="submit"
                  >
                    Login
                    <i
                      class="arrow circle right icon"
                    />
                  </button>
                </form>
              </div>
            </div>
            <div
              class="column"
            >
              <h1>
                Register
              </h1>
              <div>
                <form
                  class="ui equal width form "
                >
                  <div
                    class="fields"
                  >
                    <div
                      class="field required"
                    >
                      <label>
                        Full name
                      </label>
                      <input
                        name="full_name"
                        placeholder="First and Last Name"
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      class="field required"
                    >
                      <label>
                        Email Address
                      </label>
                      <input
                        name="email"
                        placeholder="Email Address"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="fields"
                  >
                    <div
                      class="field required"
                    >
                      <label>
                        Password
                      </label>
                      <input
                        data-testid="loginpwd1"
                        name="password"
                        type="password"
                        value=""
                      />
                    </div>
                    <div
                      class="field required"
                    >
                      <label>
                        Re-enter password
                      </label>
                      <input
                        data-testid="loginpwd2"
                        name="password2"
                        type="password"
                        value=""
                      />
                    </div>
                  </div>
                  <button
                    class="ui blue primary right floated right labeled icon button"
                    type="submit"
                  >
                    Register
                    <i
                      class="arrow circle right icon"
                    />
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ui vertical divider"
        >
          Or
        </div>
      </div>
    </div>
    <h2
      class="ui block header"
    >
      <i
        class="tasks icon"
      />
      <div
        class="content"
      >
        Browse exams
      </div>
    </h2>
    <div>
      <h3>
        No exams are available
      </h3>
    </div>
  </div>
</DocumentFragment>
`;
